gems/pending/Scvmm/miq_hyperv_disk.rb

```diff
@@ -99,15 +99,28 @@ def read(size)
   def bread_cached(start_sector, number_sectors)
     $log.debug "miq_hyperv_disk.bread_cached(#{start_sector}, #{number_sectors})"
     @block_cache.keys.each do |block_range|
-      next unless block_range.include?(start_sector) && block_range.include?(start_sector + number_sectors - 1)
       sector_offset = start_sector - block_range.first
       buffer_offset = sector_offset * @block_size
-      length        = number_sectors * @block_size
-      @cache_hits[start_sector] += 1
-      return @block_cache[block_range][buffer_offset, length]
+      if block_range.include?(start_sector) && block_range.include?(start_sector + number_sectors - 1)
+        length        = number_sectors * @block_size
+        @cache_hits[start_sector] += 1
+        return @block_cache[block_range][buffer_offset, length]
+      elsif block_range.include?(start_sector)
+        sectors_in_range = block_range.last - start_sector
+        length           = sectors_in_range * @block_size
+        remaining_blocks = number_sectors - sectors_in_range
+        @cache_hits[start_sector] += 1
+        return @block_cache[block_range][buffer_offset, length] + bread_cached(block_range.last + 1, remaining_blocks)
```

Comment:
My though was that you would either have an entire string cached, or multiple parts (usually 2) of a string cached, or one or more part(s) of a string cached and the other part(s) of the string read from disk.  In the cases where the string is concatenated from two (or more) substrings - I get that you are combining two or more substrings into one - I guess you're saying that this requires reallocating that combined string every time?  Possibly (doubly) caching the combined strings would be of help (i.e. each block is cached, and each combined block is also cached).